### PR TITLE
feat: add requires_file attr for py_wheel

### PR DESCRIPTION
**Pull request by weixiao-huang**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by pstradomski** [comment]
Any background for why this is needed? Maybe the documentation should be updated to reflect that and what's the interaction with "requires" attribute?

**Comment by weixiao-huang** [comment]
Because sometimes we may write codes like this in `setup.py`

```python
#!/usr/bin/env python

from setuptools import setup, find_packages

with open("requirements.txt") as fp:
    install_requires = fp.read()


setup(
    # get requires from file
    install_requires=install_requires,
)
```

However, in `py_wheel`, it only provides requires from bazel string. It seems we cannot get requires from files. So I add `requires_file` in `py_wheel`.

If `requires` specified, `requires_file` will not work

**Comment by weixiao-huang** [comment]
Is there any update of this PR?

**Comment by UebelAndre** [comment]
@weixiao-huang would you be able to rebase this and address conflicts?

File: python/packaging.bzl [hunk 1]
```diff
@@ -188,6 +188,12 @@ def _py_wheel_impl(ctx):
 
     if ctx.attr.python_requires:
         metadata_contents.append("Requires-Python: %s" % ctx.attr.python_requires)
+
+    if len(ctx.attr.requires) == 0 and ctx.attr.requires_file:
```
Comment (UebelAndre) [inline python/packaging.bzl:192]: Would it be possible to make this additive?

File: tools/wheelmaker.py [hunk 2]
```diff
@@ -258,6 +258,12 @@ def parse_args() -> argparse.Namespace:
     metadata_group.add_argument(
         "--platform", type=str, default="any", help="Target platform. "
     )
+    metadata_group.add_argument(
+        "--requires_file",
+        type=Path,
+        default="",
```
Comment (UebelAndre) [inline tools/wheelmaker.py:264]: I don't think there's a need to set a default here.

File: tools/wheelmaker.py [hunk 3]
```diff
@@ -398,6 +404,9 @@ def main() -> None:
                       encoding="utf-8") as metadata_file:
                 metadata = metadata_file.read()
 
+        if arguments.requires_file:
+            with open(arguments.requires_file) as fp:
```
Comment (UebelAndre) [inline tools/wheelmaker.py:408]: Shouldn't this instead be written to `Requires-Dist:` in the metadata? My use case for this is to be able to define 3rdparty dependencies in `requirements.in` files and be able to continue to use the existing `requires` field for local dependencies.

**Review by UebelAndre — CHANGES_REQUESTED** [review]
I'm no maintainer but did have some requests 😅

File: tools/wheelmaker.py [hunk 4]
```diff
@@ -398,6 +403,10 @@ def main() -> None:
...
+                additive_requires_list = ["Requires-Dist: {}".format(line) for line in fp.read().strip().split("\n")]
+                metadata += "\n" + "\n".join(additive_requires_list)
```
Comment (pstradomski) [inline tools/wheelmaker.py:409]: Won't that leave you with two \n in a row, so everything below will be interpreted as description?

Can we have a test?

File: tools/wheelmaker.py [hunk 5]
```diff
@@ -398,6 +403,10 @@ def main() -> None:
                       encoding="utf-8") as metadata_file:
                 metadata = metadata_file.read()
 
+        if arguments.requires_file:
+            with open(arguments.requires_file) as fp:
+                additive_requires_list = ["Requires-Dist: {}".format(line) for line in fp.read().strip().split("\n")]
```
Comment (pstradomski) [inline tools/wheelmaker.py:408]: why read().strip().split() and not readlines()?

**Comment by chrislovecnm** [comment]
@weixiao-huang knudge

**Comment by weixiao-huang** [comment]
Sorry, I'll check it recently

**Pull request closed without merging**